Code/EditorPlugins/VisualScript/EditorPluginVisualScript/VisualScriptGraph/VisualScriptGraph.cpp

```diff
@@ -448,18 +449,84 @@ void ezVisualScriptNodeManager::InternalCreatePins(const ezDocumentObject* pObje
   }
 }
 
-void ezVisualScriptNodeManager::GetCreateableTypes(ezHybridArray<const ezRTTI*, 32>& Types) const
+void ezVisualScriptNodeManager::GetNodeCreationTemplates(ezDynamicArray<ezNodeCreationTemplate>& out_templates) const
 {
+  auto pRegistry = ezVisualScriptNodeRegistry::GetSingleton();
+  auto propertyValues = pRegistry->GetPropertyValues();
   ezHashedString sBaseClass = GetScriptBaseClass();
 
-  for (auto it : ezVisualScriptNodeRegistry::GetSingleton()->GetAllNodeTypes())
+  for (auto& nodeTemplate : pRegistry->GetNodeCreationTemplates())
   {
-    if (IsFilteredByBaseClass(it.Key(), it.Value(), sBaseClass))
+    const ezRTTI* pNodeType = nodeTemplate.m_pType;
+
+    if (IsFilteredByBaseClass(pNodeType, *pRegistry->GetNodeDescForType(pNodeType), sBaseClass))
       continue;
 
-    if (!it.Key()->GetTypeFlags().IsSet(ezTypeFlags::Abstract))
+    if (!pNodeType->GetTypeFlags().IsSet(ezTypeFlags::Abstract))
     {
-      Types.PushBack(it.Key());
+      auto& temp = out_templates.ExpandAndGetRef();
+      temp.m_pType = pNodeType;
+      temp.m_sTypeName = nodeTemplate.m_sTypeName;
+      temp.m_sCategory = nodeTemplate.m_sCategory;
+      temp.m_PropertyValues = propertyValues.GetSubArray(nodeTemplate.m_uiPropertyValuesStart, nodeTemplate.m_uiPropertyValuesCount);
+    }
+  }
+
+  // Getter and setter templates for variables
+  if (GetRootObject()->GetChildren().IsEmpty() == false)
+  {
+    static ezHashedString sVariables = ezMakeHashedString("Variables");
```

Comment:
is it really necessary to use statics here ?

Reply:
This function is called every time the user opens the context menu to create new nodes so it better be fast.